charon/tests/layout.rs

```diff
@@ -111,8 +111,78 @@ fn type_layout() -> anyhow::Result<()> {
             B(u32) = 43,
             C = 123456,
         }
+
+        #[repr(i8)]
+        enum MyOrder {
+            Less = -1,
+            Equal = 0,
+            Greater = 1,
+        }
+
+        enum WithNicheAndUninhabited {
+            First,
+            Second(!),
+            Third(NonZero<u32>)
+        }
         "#,
     )?;
+
+    // Check whether niche discriminant computations are correct, i.e. reversible.
+    for tdecl in crate_data.type_decls.iter() {
+        if let Some(layout) = tdecl.layout.as_ref() {
+            if layout.discriminant_layout.is_some() {
+                let name = repr_name(&crate_data, &tdecl.item_meta.name);
+                let variants_from_kind = match &tdecl.kind {
+                    TypeDeclKind::Enum(variants) => variants,
+                    _ => unreachable!(),
+                };
+                for (var_id, variant) in layout.variant_layouts.iter_indexed() {
+                    let discr = variants_from_kind.get(var_id).unwrap().discriminant;
+
+                    // As discussed in https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp/topic/.E2.9C.94.20VariantId.3DDiscriminant.20when.20tag.20is.20niche.20encoded.3F/with/524384295
+                    // for niche-encoded tags, the variant id's should coincide with their discriminants.
```

Comment:
That's an implementation detail of today's rustc; rust does not guarantee this. So I'd prefer we don't test this, because if this ever breaks it won't be our fault and we don't even want to fix it.

Reply:
Funnily enough, there is a similar check in the code-gen_ssa in rustc. Nonetheless, I agree that we can remove this check.

Reply:
For reference, it's here: https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_ssa/src/mir/operand.rs#L485